torchtune/models/gemma/_tokenizer.py

```diff
@@ -0,0 +1,129 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Tuple
+
+from torchtune.data import Message
+from torchtune.modules.tokenizers import (
+    ModelTokenizer,
+    SentencePieceBaseTokenizer,
+    tokenize_messages_no_special_tokens,
+)
+
+WHITESPACE_CHARS = [" ", "\n", "\t", "\r", "\v"]
+
+
+class GemmaTokenizer(ModelTokenizer):
+    """
+    Gemma's implementation of the SentencePiece tokenizer
+
+    Args:
+        path (str): Path to pretrained tokenizer file.
+
+    Example:
+        # Accepts only non-batched input for now
+        >>> tokenizer = GemmaTokenizer("/path/to/spm_model")
+        >>> tokenized_text = tokenizer.encode("Hello world!", add_bos=True, add_eos=True)
+        >>> print(tokenized_text)
+        [1, 31587, 29644, 102, 2]
+    """
+
+    def __init__(
+        self,
+        path: str,
+    ):
+        self._spm_model = SentencePieceBaseTokenizer(path)
+
+        # Original tokenizer has no pad_id, which causes indexing errors when batch training
+        self._spm_model.pad_id = 0
+
+        # This is used in tokenize_messages: if the tokenizer does not
+        # encode whitespace, then we can more easily split strings
+        # on whitespace characters and encode them separately.
+        self.encodes_whitespace = any(
+            [self._spm_model.encode(c) for c in WHITESPACE_CHARS]
+        )
+
+        # During generation, stop when eos_id is encountered
+        self.stop_tokens = [self.eos_id]
+
+    @property
+    def eos_id(self):
+        return self._spm_model.eos_id
+
+    @property
+    def bos_id(self):
+        return self._spm_model.bos_id
+
+    @property
+    def pad_id(self):
+        return self._spm_model.pad_id
+
+    @property
+    def vocab_size(self):
+        return self._spm_model.vocab_size
+
+    def encode(
```

Comment:
Discussed a bit offline, but currently the encode/decode methods being defined in both Tokenizer and ModelTokenizer is confusing. Right now it's needed because we do call encode directly on ModelTokenizer in our generate recipe, but as @RdoubleA pointed out we should see if we can move everything onto tokenize_messages API (and deprecate encode/decode methods for ModelTokenizers) after #1106